Tracing: export opentelemetry metrics with span-level statistics

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/17325?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17325/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17325/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17325/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Tracing: export opentelemetry metrics with span-level statistics

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [X] Examples
  - [ ] API references
  - [X] Instructions

### Release Notes

Added support for recording statistics about traces, such as duration & number of spans logged, to OpenTelemetry

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [X] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

dbczumar committed:
Add comprehensive test for dual export to both MLflow and OTLP

- Test verifies traces are sent to both OTLP collector and MLflow backend
- Uses synchronous MLflow export with MLFLOW_ENABLE_ASYNC_TRACE_LOGGING=false
- Minimal assertions to verify spans reach both destinations
- Test currently reveals dual export bug: traces only go to OTLP, not MLflow

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix dual export bug where traces weren't sent to MLflow backend

When OTLP dual export mode is enabled, both OTLP and MLflow processors
are active. Previously, the OTLP processor was managing the trace lifecycle
(registering and popping traces from the manager) even in dual mode, which
prevented the MLflow processor from accessing the traces.

This fix ensures that in dual export mode:
- The OTLP processor only exports spans to OTLP collector
- The MLflow processor handles all trace lifecycle management
- Both backends receive the trace data correctly

The bug was causing traces to only be sent to OTLP and not MLflow when
dual export was enabled.

Signed-off-by: corey.zumar <corey.zumar@databricks.com>
Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add MLFLOW_LOG_OTLP_TRACE_STATISTICS env var for OTLP trace metrics

- Add environment variable to enable logging of trace statistics via OTLP metrics
- Update OtelSpanProcessor to emit histogram metrics when spans end
- Metrics include labels for root span status, span type, and span status
- Records span duration in seconds to mlflow.trace.span.duration histogram

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Change span duration metric from seconds to milliseconds

More practical unit for measuring span durations

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge otel_dual_export branch and add OTLP trace statistics

- Merge MLFLOW_TRACE_ENABLE_OTLP_DUAL_EXPORT functionality
- Add MLFLOW_LOG_OTLP_TRACE_STATISTICS for metrics collection
- Integrate trace registration with OTLP metrics logging

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Raise exception for unrecognized OTLP metrics protocols

- Add explicit MlflowException for unsupported protocols
- Document valid protocols per OpenTelemetry spec
- Improve error message with supported protocol list

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge remote-tracking branch 'coreyssh/otel_dual_export' into log_otel_metric

dbczumar committed:
merge

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add OpenTelemetry metrics export support

- Add export_spans and export_metrics parameters to OtelSpanProcessor
- Enable metrics-only export when OTEL_EXPORTER_OTLP_METRICS_ENDPOINT is set
- Remove MLFLOW_LOG_OTLP_TRACE_STATISTICS dependency for simplified configuration
- Update documentation with metrics export setup instructions
- Add should_export_otlp_metrics() helper function for endpoint detection
- Support gRPC and HTTP protocols for metrics export
- Export mlflow.trace.span.duration histogram with span metadata

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove default values from OtelSpanProcessor export parameters

Make export_spans and export_metrics required parameters to ensure
explicit configuration at instantiation time.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor _get_experiment_id_for_trace to reusable utility

- Move _get_experiment_id_for_trace from BaseMlflowSpanProcessor to tracing utils
- Rename to get_experiment_id_for_trace for general use across exporters
- Update MlflowV3SpanProcessor to use the new utility function
- Remove unused imports from base_mlflow.py
- Make experiment ID resolution logic reusable across all span processors

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Integrate experiment ID into OtelSpanProcessor and metrics

- Add experiment ID to _create_trace_info using get_experiment_id_for_trace
- Include experiment_id attribute in duration histogram metrics
- Update documentation to reflect experiment_id label in exported metrics
- Enable performance analysis grouped by MLflow experiment

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Enhance metrics with trace tags and metadata

- Include trace tags as 'tag_*' attributes in duration histogram (root spans only)
- Add select trace metadata as 'meta_*' attributes (source_run, model_id)
- Sanitize tag keys for metric label compatibility (replace . and - with _)
- Update documentation to reflect enhanced metric attributes
- Enable richer performance analysis by trace characteristics

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Include trace tags and metadata for ALL spans, not just root spans

- Remove is_root condition for trace tags and metadata inclusion
- All spans now get trace-level context in their metrics
- Update documentation to reflect tags/metadata available for all spans
- Enable richer analysis across all span levels in a trace

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Always fetch trace data for metrics regardless of trace registration

- Remove _should_register_traces condition when fetching trace tags/metadata
- Ensure metrics always have trace context even in metrics-only mode
- Critical for scenarios where spans are exported but traces aren't registered

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix trace data retrieval to use correct trace ID

- Use span.context.trace_id instead of REQUEST_ID attribute for get_trace()
- The trace manager operates on OpenTelemetry trace IDs, not MLflow request IDs
- Remove unused get_otel_attribute import
- Fix indentation and simplify trace data access logic

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Include ALL trace metadata in metrics, not just select fields

- Export all trace metadata fields as meta_* attributes
- Apply same key sanitization as tags (replace . and - with _)
- Update documentation to reflect complete metadata inclusion
- Maximize analytical capabilities with full trace context

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary key sanitization for metric attributes

- Remove .replace() calls for tag and metadata keys
- OpenTelemetry handles original key names just fine
- Simplify code and preserve original key names
- Update documentation examples to show actual key formats

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update documentation for new tags. and metadata. prefixes

- Change tag_* to tags.* in documentation examples
- Change meta_* to metadata.* in documentation examples
- Update metric attribute examples to reflect new naming scheme
- Maintain consistency with code changes to use dot notation

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix OtelSpanProcessor trace registration and tag/metadata export

- Always initialize trace manager for metrics (needed to access trace context)
- Register traces for both span export AND metrics export modes
- Use proper OTel-to-MLflow trace ID mapping for trace lookups
- Fix JSON decoding of span types using json.loads instead of manual parsing
- Enable tags and metadata export in metrics-only mode
- All trace tags/metadata now properly included in histogram attributes

This ensures mlflow.trace.span.duration metrics include comprehensive
trace context (tags.* and metadata.*) even in metrics-only export mode.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add direct Prometheus OTLP receiver documentation for metrics export

- Document direct Prometheus export option alongside OpenTelemetry Collector approach
- Add required Prometheus flags: --web.enable-otlp-receiver and --enable-feature=otlp-deltatocumulative
- Provide complete configuration example for direct OTLP ingestion
- Organize documentation with separate subsection for Prometheus approach

This enables simplified metrics monitoring without requiring an OpenTelemetry Collector.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix metadata examples in tracing docs to use correct built-in keys

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge remote-tracking branch 'origin/master' into otel_metric_prom

dbczumar committed:
Fix trace cleanup logic in OTLP processor

- Only pop traces when actually registering them (_should_register_traces)
- Don't pop traces when only exporting metrics (_export_metrics)
- This fixes metrics export where traces were being cleaned up prematurely
- Resolves issue where Prometheus metrics had no data due to trace ID mapping being lost

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add comprehensive test suite for OTEL metrics export functionality

This test suite validates:
- Processor initialization with metrics export enabled
- Metrics collection from real and mock spans
- Critical trace cleanup behavior (ensuring traces aren't cleaned up during metrics-only export)
- JSON decoding of span types
- Comprehensive trace attributes in metrics (tags, metadata, experiment ID)
- Disabled metrics functionality

Tests use proper mocking to avoid network calls and follow existing MLflow testing patterns.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor OTEL metrics tests to focus on behavior over internal state

- Remove brittle assertions on internal properties (_export_metrics, etc.)
- Create reusable create_mock_span() utility to reduce code duplication
- Focus tests on observable behavior: method calls, histogram recording, trace cleanup
- Test what the code DOES rather than how it's internally structured
- Maintain same comprehensive test coverage with better encapsulation

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Clean up OTEL metrics tests: remove sleeps and useless comments

- Remove time.sleep() calls that were unnecessary and slowing down tests
- Strip out verbose inline comments that added no value
- Keep tests focused on essential behavior verification
- Tests now run ~20% faster with same comprehensive coverage

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify provider logic and add comprehensive OTLP test coverage

- Remove redundant should_export_spans check in provider.py
- Add 3 new test cases in provider test suite:
  * test_otlp_metrics_only_export: verifies metrics-only with NoOpSpanExporter
  * test_otlp_spans_and_metrics_export: verifies both spans and metrics
  * test_otlp_no_export: verifies no OtelSpanProcessor when neither enabled
- Tests focus on observable behavior rather than internal state
- Provides comprehensive coverage of all OTLP export combinations

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix provider tests to use environment variables instead of mocking

- Replace mock.patch() calls on should_use_otlp_exporter/should_export_otlp_metrics
- Use actual OTEL environment variables to test real behavior:
  * OTEL_EXPORTER_OTLP_TRACES_ENDPOINT for span export
  * OTEL_EXPORTER_OTLP_METRICS_ENDPOINT for metrics export
  * OTEL_EXPORTER_OTLP_*_PROTOCOL for transport configuration
- Tests now validate actual environment-driven behavior instead of mocked functions
- More robust and realistic test coverage

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove inline comments from OTEL test files

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Merge remote-tracking branch 'origin/master' into otel_metric_prom

dbczumar committed:
Add OtelMetricsMixin to share metrics recording logic between span processors

- Created OtelMetricsMixin class to encapsulate OpenTelemetry metrics recording functionality
- Integrated mixin into OtelSpanProcessor and BaseMlflowSpanProcessor
- Removed duplicate metrics code from OtelSpanProcessor
- Added export_metrics parameter to control metrics export behavior

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update tracing provider to support metrics export configuration

- Modified provider.py to handle export_metrics parameter
- Only export metrics from OtelSpanProcessor when dual export is disabled to avoid duplication

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix tests for OtelMetricsMixin integration

- Updated test_otel_metrics.py to reflect new lazy initialization of metrics
- Removed obsolete test_metrics_only_no_trace_cleanup test case
- Fixed test signatures to remove obsolete export_spans parameter
- Removed invalid test_otlp_metrics_only_export test case from provider tests
  (metrics without spans doesn't make logical sense)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Change unsupported OTLP protocol handling from exception to warning

- Log warning and skip metrics export for unsupported protocols instead of throwing exception
- This prevents failures when encountering unknown protocol configurations

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix test_mlflow_v3_processor.py tests for new export_metrics parameter

- Added export_metrics=False parameter to all MlflowV3SpanProcessor instantiations in tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add opentelemetry-exporter-otlp requirement to docs and skip test in SDK-only mode

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add opentelemetry-exporter-otlp requirement to docs and skip test in SDK-only mode

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix Windows CI test failure by ensuring measurable span duration

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/tracing/processor/base_mlflow.py:
```diff
@@ -100,31 +105,6 @@ def on_end(self, span: OTelReadableSpan) -> None:
 
         super().on_end(span)
 
-    def _get_experiment_id_for_trace(self, span: OTelReadableSpan) -> str:
-        """
-        Determine the experiment ID to associate with the trace.
-
-        The experiment ID can be configured in multiple ways, in order of precedence:
-          1. An experiment ID specified via the span creation API i.e. MlflowClient().start_trace()
-          2. An experiment ID specified via `mlflow.tracing.set_destination`
-          3. An experiment ID of an active run.
-          4. The default experiment ID
-        """
-        from mlflow.tracing.provider import _MLFLOW_TRACE_USER_DESTINATION
-        from mlflow.tracking.fluent import _get_latest_active_run
-
-        if experiment_id := get_otel_attribute(span, SpanAttributeKey.EXPERIMENT_ID):
-            return experiment_id
-
-        if destination := _MLFLOW_TRACE_USER_DESTINATION.get():
-            if exp_id := getattr(destination, "experiment_id"):
-                return exp_id
-
-        if run := _get_latest_active_run():
-            return run.info.experiment_id
-
-        return _get_experiment_id()
```
This is needed in OtelMetricsMixin and MlflowV3SpanProcessor; I've moved it out as a reusable utility method, since I think it'll be helpful in the future beyond these two classes

dbczumar committed:
Simplify status code conversion using enum name property

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify should_export_otlp_metrics function

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary try/catch for OpenTelemetry SDK imports

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify OTLP metrics export test to focus on essential verification

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary test_otlp_no_export test case

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Refactor metrics tests to use InMemoryMetricReader instead of mocking

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify test assertions to directly check each expected metric

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use different sleep durations for each span type in tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use very different sleep durations (10ms, 250ms, 1s) to avoid test flakiness

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary OtelSpanProcessor instantiation and sort metrics by duration

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify assertion for missing metrics in disabled test

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Clean up test imports and simplify processor checks

- Move imports to top of test file
- Remove unnecessary complexity in processor checks (always wrapped in SynchronousMultiSpanProcessor)

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use explicit 'is not None' pattern for clarity

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unnecessary intermediate variables for cleaner code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use keyword argument for histogram record amount parameter

Signed-off-by: dbczumar <corey.zumar@databricks.com>

harupy commented:
/review

harupy commented on mlflow/tracing/utils/__init__.py:
```diff
@@ -545,6 +545,38 @@ def update_trace_state_from_span_conditionally(trace, root_span):
+        return experiment_id
+
+    if destination := _MLFLOW_TRACE_USER_DESTINATION.get():
+        if exp_id := getattr(destination, "experiment_id"):
```
```suggestion
        if exp_id := getattr(destination, "experiment_id", None):
```

dbczumar replied:
Fixed!

harupy commented on tests/tracing/processor/test_otel_metrics.py:
```diff
@@ -0,0 +1,108 @@
+                    assert metric.unit == "ms"
+                    data_points.extend(metric.data.data_points)
+
+    assert len(data_points) == 3, "Expected exactly 3 span metrics"
```
```suggestion
    assert len(data_points) == 3
```

this seems redundant.

harupy commented on mlflow/tracing/processor/otel_metrics_mixin.py:
```diff
@@ -0,0 +1,127 @@
+        """
+        self._setup_metrics_if_necessary()
+
+        if not self._duration_histogram:
```
```suggestion
        if not self._duration_histogram is not None:
```

for consistency

dbczumar replied:
Fixed!

harupy commented on mlflow/tracing/processor/otel_metrics_mixin.py:
```diff
@@ -0,0 +1,127 @@
+        if self._duration_histogram is not None:
+            return
+
+        endpoint = os.environ.get("OTEL_EXPORTER_OTLP_METRICS_ENDPOINT") or os.environ.get(
```
can use `_get_otlp_metrics_endpoint` here?

dbczumar replied:
Absolutely! Fixed!

dbczumar committed:
Address PR review comments from @harupy

- Add None default to getattr in utils/__init__.py
- Remove assertion message in test_otel_metrics.py
- Use existing helper functions from utils/otlp.py
- Add default_value parameter to protocol helper functions

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix test_incremental_span_name_deduplication - add missing export_metrics parameter

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/tracing/processor/otel.py:
```diff
@@ -6,20 +6,22 @@
     """
 
-    def __init__(self, span_exporter: SpanExporter):
+    def __init__(self, span_exporter: SpanExporter, export_metrics: bool):
```
that would be error prone; I'd rather be explicit. All it takes is one call site unintentionally failing to specify a value to cause a problem.

dbczumar replied:
Yeah, it's only used in MLflow

harupy commented on mlflow/tracing/processor/otel_metrics_mixin.py:
```diff
@@ -0,0 +1,123 @@
+            pass
+
+        attributes = {
+            "root": str(span.parent is None),
```
Ton confirm, this is correct?

dbczumar replied:
Yeah, if a span doesn't have a parent, it's the root

harupy replied:
`attributes` only accept strings?

dbczumar replied:
Ah, fair point - we don't technically have to stringify (boolean is supported). I'll remove str.

harupy reviewed (approved):
LGTM once https://github.com/mlflow/mlflow/pull/17325/files#r2312817125 is answered/resolved!

dbczumar committed:
Address all Codex-generated PR review comments

- Add return type hints to __init__ methods
- Add type hints to test functions and fixtures
- Remove redundant test docstrings that restate function names
- Add proper typing imports where needed

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use boolean instead of string for 'root' metric attribute

Signed-off-by: dbczumar <corey.zumar@databricks.com>
